trackintel/analysis/location_identification.py

```diff
@@ -229,3 +233,81 @@ def _freq_assign(duration, *labels):
     label_array = np.full(len(duration), fill_value=None)
     label_array[kth] = labels
     return label_array
+
+
+def osna_method(spts):
+    """Use weekdays data divided in three time frames ["rest", "work", "leisure"]. Finds most popular
+    home location for timeframes "rest" and "leisure" and most popular "work" location for "work" timeframe.
+
+    "rest" + "leisure" durations are weighted together. Location with longest duration is assigned "home" label.
+    "work" locations with longest duration is assigned "work" label.
+    The "weekend" label isn't used for any label.
+
+    Parameters
+    ----------
+    spts : GeoDataFrame (as trackintel staypoints)
+        Staypoints with the column "location_id"
+
+    Returns
+    -------
+    GeoDataFrame
+        with additional column "activity_label"
+
+    Note
+    ----
+    The method is adapted from [1].
+    When "home" and "work" label overlap, the method selects the next location by 2nd highest score.
+    The original algorithm count the distinct hours at a location as the home location is derived from
+    geo-tagged tweets. We directly sum the time spent at a location as our data model includes that.
+
+    References
+    ----------
+    [1] Efstathiades, Hariton, Demetris Antoniades, George Pallis, and Marios Dikaiakos. 2015.
+    ‘Identification of Key Locations Based on Online Social Network Activity’.
+    In https://doi.org/10.1145/2808797.2808877.
+
+    Examples
+    --------
+    >>> staypoints = ti.analysis.osna_method(staypoints)
+
+    """
+    spts_in = spts  # not sure how to handle changing columns
+    spts = spts.copy()
+    spts["duration"] = spts["finished_at"] - spts["started_at"]
+    spts["mean_time"] = spts["started_at"] + spts["duration"] / 2
+
+    spts["label"] = spts["mean_time"].apply(_osna_label_timeframes)
+    spts.loc[spts["label"] == "rest", "duration"] *= 0.739  # weight given in paper
+    spts.loc[spts["label"] == "leisure", "duration"] *= 0.358  # weight given in paper
+    groups_map = {
+        "rest": "home",
+        "leisure": "home",
+        "work": "work",
+    }  # weekends aren't included in analysis
+    groups = ["user_id", "location_id", spts["label"].map(groups_map)]
+    spts_pivot = spts.groupby(groups)["duration"].sum().unstack()  # pivot table with "label" in columns
+    spts_idxmax = spts_pivot.groupby(["user_id"]).idxmax()  # maximum for home and work
+    spts_pivot.loc[spts_idxmax["home"], "activity_label"] = "home"
+    # The "home" label could overlap with the "work" label
+    # we set the rows where that happens to zero and recalculate work maximum.
+    # alternatively we could assign activity label on staypoint level instead of location level?
```

Comment:
I'm not sure about that either. I mean, it's possible to just aggregate everything at the location level and then decide in what time frame the mean time of the location is. But then we have the risk of a location being boosted by staypoints that are not in that time frame at all.
On the other side the current way is kind of a hack, because we can only assign a label to the location and then have to handle that somehow.